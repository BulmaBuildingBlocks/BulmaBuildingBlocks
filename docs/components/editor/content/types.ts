export enum EditableContentTypes {
  Text = 'text',
  Title = 'title'
}

export interface EditableContentProps {
  /**
   * Element wrapping the Editable Content.
   */
  tag?: string;

  /**
   * Provides different editing options
   */
  type?: EditableContentTypes;

  /**
   * Class placed on the parent element.
   */
  class?: string;

  /**
   * Value that's edited.
   */
  value: string;

  /**
   * Content Editable.
   */
  editable?: boolean;
}
