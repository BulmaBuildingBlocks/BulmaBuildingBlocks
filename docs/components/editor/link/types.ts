import { Status } from '~/shared/config';

export enum LinkType {
  Link = 'Link',
  Button = 'button',
  Rounded = 'button is-rounded',
  Outlined = 'button is-outlined',
  'Rounded Outlined' = 'button is-rounded is-outlined'
}

export class EditableLink {
  constructor(props: EditableLinkProps) {
    this.label = props.label;
    this.status = props.status || Status.None;
    this.linkType = props.linkType;
    this.href = props.href;
    this.popupPlacement = props.popupPlacement;
  }

  /*
   * Label on the button.
   */
  label: string;

  /**
   * Assigned hyperlink.
   */
  href?: string;

  /**
   * Link styles.
   */
  linkType: LinkType;

  /**
   * Link styles.
   */
  status: Status;

  /**
   * Popup Placement.
   */
  popupPlacement?: string;
}

export interface EditableLinkProps {
  /**
   * Label on the button.
   */
  label: string;

  /**
   * Assigned hyperlink.
   */
  href?: string;

  /**
   * Link styles.
   */
  linkType: LinkType;

  /**
   * Link styles.
   */
  status?: Status;

  /**
   * Popup Placement.
   */
  popupPlacement?: string;
}
