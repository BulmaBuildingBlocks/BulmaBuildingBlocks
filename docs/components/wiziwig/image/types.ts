export interface EditableImageProps {
  /**
   * Key of a permalink referencing the source of the picture.
   */
  src: string;

  /**
   * Caption on the picture, used also as alternative text.
   */
  alt: string;

  /**
   * Hyperlink attached to the picture.
   */
  href?: string;

  /**
   * Picture width.
   */
  width?: number;

  /**
   * Picture height.
   */
  height?: number;

  /**
   * Popup Placement.
   */
  popupPlacement?: string;
}
