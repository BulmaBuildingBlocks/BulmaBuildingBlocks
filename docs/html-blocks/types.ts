import { VueConstructor } from 'vue';

export enum ContentTypes {
  Headers = 'headers',
  Features = 'features',
  CallToActions = 'calltoactions',
  Contents = 'contents',
  Footers = 'footers'
}

export interface Block {
  title: string;
  type?: ContentTypes;
  componentName: string;
  component: VueConstructor;
  content?: any;
  color?: string;
  container?: boolean;
}

export interface BlockListObject {
  [x: string]: VueConstructor;
}
