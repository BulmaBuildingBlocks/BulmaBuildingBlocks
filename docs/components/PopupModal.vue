<template>
  <span>
    <span v-if="show" class="popup">
      <slot />
    </span>
    <slot name="reference" />
  </span>
</template>

<script lang="ts">
import { createPopper, Instance, Options } from '@popperjs/core';
import { Component, Prop, Vue, Watch } from 'nuxt-property-decorator';
import { VNode } from 'vue';

@Component
export default class PopupModal extends Vue {
  @Prop({ type: Object, default: {} }) options!: object;
  @Prop({ default: () => false }) show!: boolean;

  popperInstance?: Instance = undefined;
  popperOptions: Options = {
    strategy: 'absolute',
    placement: 'top',
    modifiers: [
      {
        name: 'offset',
        options: {
          offset: [0, 10]
        }
      },
      {
        name: 'preventOverflow',
        options: {
          padding: 5
        }
      },
      {
        name: 'flip',
        options: {
          fallbackPlacements: ['bottom', 'right', 'left']
        }
      },
      {
        name: 'computeStyles',
        options: {
          gpuAcceleration: false
        }
      }
    ]
  };

  referenceElm?: HTMLElement = undefined;
  content?: HTMLElement = undefined;

  $scopedSlots!: {
    reference: () => VNode[];
    default: () => VNode[];
  };

  @Watch('show', { immediate: true })
  async onShowPopperChange(show: boolean) {
    if (show) {
      await this.$nextTick();
      this.showPopup();
    } else {
      this.hidePopup();
    }
  }

  created() {
    this.popperOptions = Object.assign(this.popperOptions, this.options);
  }

  async showPopup() {
    this.referenceElm = (await this.$scopedSlots.reference()[0]
      .elm) as HTMLElement;
    this.content = (await this.$scopedSlots.default()[0].elm) as HTMLElement;

    if (this.referenceElm && this.content) {
      this.popperInstance = createPopper(
        this.referenceElm,
        this.content,
        this.popperOptions
      );

      await this.popperInstance.update();
    }
  }

  hidePopup() {
    if (this.popperInstance) {
      this.popperInstance.destroy();
      this.popperInstance = undefined;
    }
  }
}
</script>
