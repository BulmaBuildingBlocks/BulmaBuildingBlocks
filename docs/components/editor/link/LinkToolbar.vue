<template>
  <div class="image-toolbar card">
    <header class="card-header">
      <p class="card-header-title">
        Configuration
      </p>
      <a class="card-header-icon" aria-label="close" @click="closeModal">
        <b-icon pack="fa" icon="times" />
      </a>
    </header>
    <div class="card-content">
      <b-field label="Label" placeholder="http://">
        <b-input v-model="newValue.label" />
      </b-field>
      <b-field label="Link">
        <b-input v-model="newValue.href" />
      </b-field>
      <b-field label="Type">
        <b-select v-model="newValue.linkType">
          <option v-for="(option, key) in LinkType" :key="key" :value="option">
            {{ key }}
          </option>
        </b-select>
      </b-field>

      <b-field label="Color">
        <color-picker
          :color="newValue.status"
          @color="newValue.status = $event"
        />
      </b-field>
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component, PropSync } from 'nuxt-property-decorator';
import { EditableLinkProps, LinkType } from '~/components/editor/link/types';
import ColorPicker from '~/components/ColorPicker.vue';

/***
 * Editing Features:
 * Upload and replace image
 * Alt tag
 * Link
 */

@Component({
  components: {
    ColorPicker
  }
})
export default class LinkToolbar extends Vue {
  @PropSync('value', { type: Object }) newValue!: EditableLinkProps;

  LinkType = LinkType;

  closeModal() {
    this.$emit('closeModal');
  }
}
</script>
