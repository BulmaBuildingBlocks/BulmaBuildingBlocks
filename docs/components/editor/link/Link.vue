<script lang="ts">
import { CreateElement } from 'vue';
import { Component, Prop, PropSync, Vue } from 'nuxt-property-decorator';
import ButtonToolbar from './LinkToolbar.vue';
import PopupModal from '~/components/PopupModal.vue';

import { EditableLinkProps, LinkType } from '~/components/editor/link/types';
import ClickOutside from '~/directives/click-outside';

/***
 * Editting Features:
 * Upload and replace link
 * Alt tag
 * Link
 */

@Component({
  directives: {
    ClickOutside
  }
})
export default class Link extends Vue {
  @Prop(Boolean) editable!: boolean;
  @PropSync('value', { type: Object }) newValue!: EditableLinkProps;

  modalOpen = false;

  get linkClasses() {
    const linkType = this.newValue.linkType.toLowerCase();
    let linkStyle = '';

    if (this.newValue.status) {
      if (this.newValue.linkType === LinkType.Link) {
        linkStyle = `has-text-${this.newValue.status.toLowerCase()}`;
      } else {
        linkStyle = `is-${this.newValue.status?.toLowerCase()}`;
      }
    }

    return [linkType, linkStyle];
  }

  toggleModal(event: Event) {
    event.preventDefault();

    this.modalOpen = !this.modalOpen;
  }

  closeModal() {
    this.modalOpen = false;
  }

  render(createElement: CreateElement) {
    const link = createElement(
      'a',
      {
        class: this.linkClasses,
        on: {
          click: this.toggleModal
        },
        attrs: {
          href: this.newValue.href
        }
      },
      this.newValue.label
    );

    const linkToolbar = createElement(ButtonToolbar, {
      props: {
        value: this.newValue
      },
      on: {
        click: this.toggleModal,
        closeModal: () => {
          this.closeModal();
        },
        'update:value': (value: EditableLinkProps) => {
          this.newValue = value;
        }
      }
    });

    return createElement(PopupModal, {
      scopedSlots: {
        reference: () => {
          return link;
        },
        default: () => {
          return linkToolbar;
        }
      },
      props: {
        show: this.editable && this.modalOpen,
        options: {
          placement: this.newValue.popupPlacement || 'right-start'
        }
      },
      directives: [
        {
          name: 'click-outside',
          value: { handler: this.closeModal }
        }
      ]
    });
  }
}
</script>
