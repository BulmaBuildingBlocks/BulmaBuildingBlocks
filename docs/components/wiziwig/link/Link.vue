<script lang="ts">
import { CreateElement } from 'vue';
import { Vue, Component, Prop, PropSync } from 'nuxt-property-decorator';
import ButtonToolbar from './LinkToolbar.vue';
import PopupModal from '~/components/global/PopupModal.vue';

import { EditableLinkProps } from '~/components/wiziwig/link/types';
import ClickOutside from '~/directives/click-outside';

/***
 * Editting Features:
 * Upload and replace button
 * Alt tag
 * Link
 */

@Component({
  directives: {
    ClickOutside
  }
})
export default class Link extends Vue {
  @Prop(Boolean) editable!: boolean;
  @PropSync('value', { type: Object }) newValue!: EditableLinkProps;

  modalOpen = false;

  toggleModal(event: Event) {
    event.preventDefault();

    this.modalOpen = !this.modalOpen;
  }

  openModal(event: Event) {
    event.preventDefault();

    this.modalOpen = true;
  }

  closeModal() {
    this.modalOpen = false;
  }

  render(createElement: CreateElement) {
    const button = createElement(
      'a',
      {
        class: this.newValue.styles,
        on: {
          click: this.toggleModal
        },
        attrs: {
          href: this.newValue.href
        }
      },
      this.newValue.label
    );

    const buttonToolbar = createElement(ButtonToolbar, {
      props: {
        value: this.newValue
      },
      on: {
        click: this.toggleModal,
        closeModal: () => {
          this.closeModal();
        },
        'update:value': (value: EditableLinkProps) => {
          this.newValue = value;
        }
      }
    });

    const popupModal = createElement(PopupModal, {
      scopedSlots: {
        reference: () => {
          return button;
        },
        default: () => {
          return buttonToolbar;
        }
      },
      props: {
        show: this.editable && this.modalOpen,
        options: {
          placement: this.newValue.popupPlacement || 'right-start'
        }
      },
      directives: [
        {
          name: 'click-outside',
          value: { handler: this.closeModal }
        }
      ]
    });

    return popupModal;
  }
}
</script>
