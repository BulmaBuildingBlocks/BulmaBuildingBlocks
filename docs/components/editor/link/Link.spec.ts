import { mount, Wrapper } from '@vue/test-utils';
import ButtonToolbar from './LinkToolbar.vue';
import Link from '~/components/editor/link/Link.vue';
import { EditableLinkProps, LinkType } from '~/components/editor/link/types';
import { Status } from '~/shared/config';
import PopupModal from '~/components/PopupModal.vue';
import ClickOutside from '~/directives/click-outside';

const createWrapper = (propsData: EditableLinkProps) => {
  return mount(Link, {
    propsData: { value: propsData, editable: true },
    stubs: {
      transition: false, // fixes error when using transition component
      PopupModal,
      ButtonToolbar
    },
    directives: {
      ClickOutside
    }
  });
};

describe('Link.vue', () => {
  let wrapper: Wrapper<Link & { [key: string]: any }>;

  beforeEach(() => {
    // Mount the component
    wrapper = createWrapper({
      label: '',
      linkType: LinkType.Link
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('has the expected html structure', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test('is a Vue instance', () => {
    expect(wrapper.isVueInstance).toBeTruthy();
  });

  test('correct color class returned in linkClasses', async () => {
    wrapper = createWrapper({
      label: '',
      status: Status.None,
      linkType: LinkType.Link
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.linkClasses).not.toContain('has-text-none');
    expect(wrapper.vm.linkClasses).not.toContain('is-none');

    wrapper = createWrapper({
      label: '',
      status: Status.Primary,
      linkType: LinkType.Link
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.linkClasses).toContain('has-text-primary');

    wrapper = createWrapper({
      label: '',
      status: Status.Primary,
      linkType: LinkType.Button
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.linkClasses).toContain('is-primary');

    wrapper = createWrapper({
      label: '',
      status: Status.Dark,
      linkType: LinkType.Outlined
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.linkClasses).toContain('is-dark');
  });

  test('toggle model toggles', () => {
    wrapper.vm.toggleModal(new Event('clicked'));
    expect(wrapper.vm.modalOpen).toEqual(true);
    wrapper.vm.toggleModal(new Event('clicked'));
    expect(wrapper.vm.modalOpen).toEqual(false);
  });

  test('close modal function', () => {
    expect(typeof wrapper.vm.closeModal).toBe('function');

    wrapper.vm.closeModal();
    expect(wrapper.vm.modalOpen).toEqual(false);
  });

  test('open and close popup toolbar', async () => {
    wrapper = createWrapper({
      label: 'button text',
      status: Status.Primary,
      linkType: LinkType.Button
    });

    // Check button and check popup is shown
    await wrapper.find('.button').trigger('click');
    expect(wrapper.find('.popup').exists()).toBeTruthy();

    // Check button and check popup is hidden
    await wrapper.find('.card-header-icon').trigger('click');
    expect(wrapper.find('.popup').exists()).toBeFalsy();

    // Check button and check popup is shown
    await wrapper.find('.button').trigger('click');
    expect(wrapper.find('.popup').exists()).toBeTruthy();

    // Check button and check popup is hidden
    await wrapper.find('span > a.button').trigger('click');
    expect(wrapper.find('.popup').exists()).toBeFalsy();
  });
});
