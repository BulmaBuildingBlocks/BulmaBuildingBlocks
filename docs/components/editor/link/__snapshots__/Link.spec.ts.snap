// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link.vue has the expected html structure 1`] = `
<span>
  <!---->
   
  <a
    class="link"
  >
    
  </a>
</span>
`;
